doc/monitoring_high_mem_containers/monitoring_containers.md

```diff
@@ -0,0 +1,282 @@
+
+Specifically, for monitoring memory usage of a container, the workflow can be
+described as the following steps:
+1.  Monit spawns a process to execute the script `memory_checker` every 1 minute
+2.  `memory_checker` accepts <container_name> as a parameter and retrieves 
+    alerting status and memory threshold of this container from `CONFIG_DB`.
+3.  If high memory alerting is enabled and the container is running, `memory_checker` 
+    retrieves its runtime memory usage from command output of `docker stats`; 
+    Otherwiese, `memory_checker` exits and logs an message indicating
+    either alerting is disabled or specified container is not running.
+4.  If runtime memory usage is larger than memory threshold, then `memory_checker`
+    exits with non-zero value; Otherwise, `memory_checker` exits with zero value.
+5.  Monit will write an alerting message into syslog if it receives non-zero
+    value from `memory_checker` for specified number of times during a monitoring
+    interval.
+6.  After the monitoring interval, Monit will write alerting messages into syslog
+    every 1 minute if it receives non-zero value from `memory_checker` in every
+    1 minute polling cycle.
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+This mechanism is used to perform monitoring memory usage of docker containers:
+1.  Monit will write an alerting message into syslog if it receives non-zero
+    value from `memory_checker` for specified number of times during a monitoring
+    interval.
+    A non-zero value indicates runtime memory usage of a docker container is
+    larger than its memory threshold.
+2.  After the monitoring interval, Monit will write alerting messages into syslog
+    every 1 minute if it receives non-zero value from `memory_checker` in every
+    1 minute polling cycle.
+
+
+## 2.2 Functional Description
+
+
+### 2.2.1 Monitoring Memory Usage of Containers
+Monit can be employed to monitor the resource usage such as CPU, memory and disk
+of each process. Unfortunately Monit is unable to do the resource monitoring
+in the container level. Thus we propose a mechanism to monitor memory usage of
+docker containers based on Monit.
+
+Specifically, Monit will launch a script `memory_checker` and check its exit value.
+A non-zero value indicates runtime memory usage of a docker container is larger than
+its memory threshold. Monit will write an alerting message into syslog if it receives
+non-zero value from `memory_checker` for specified number of times during a monitoring
+interval. After the monitoring interval, Monit will write alerting messages into syslog
+every 1 minute if it receives non-zero value from `memory_checker` in every 1 minute
+polling cycle.
+
+Below is an example showing Monit configuration file of lldp container:
+
+```bash
+check program container_memory_lldp with path "/usr/bin/memory_checker lldp"
+    if status == 3 for 10 times within 20 minutes then alert repeat every 1 cycles
```

Comment:
>3 [](http://example.com/codeflow?start=17&length=1)

You did not define `3` as a non-zero value. Could you add all the possible values with its meaning?